src/bio/bio_xstream.c

```diff
@@ -367,8 +371,27 @@ bio_nvme_init(const char *nvme_conf, int shm_id, int mem_size,
 		goto free_mutex;
 	}
 
+	D_ASSERT(tgt_nr > 0);
+	D_ASSERT(mem_size > 0);
+	D_ASSERT(hugepage_size > 0);
+	/*
+	 * Hugepages are not enough to sustain average I/O workload
+	 * (~1GB per xstream).
+	 */
+	if ((mem_size / tgt_nr) < DAOS_DMA_MIN_UB_BUF_MB) {
+		D_ERROR("Per-xstream DMA buffer upper bound limit < 1GB!\n");
+		D_DEBUG(DB_MGMT, "mem_size:%dMB, DMA upper bound:%dMB\n",
+			mem_size, (mem_size / tgt_nr));
+		return -DER_INVAL;
+	}
+
 	bio_chk_cnt_init = DAOS_DMA_CHUNK_CNT_INIT;
-	bio_chk_cnt_max = DAOS_DMA_CHUNK_CNT_MAX;
+	bio_chk_cnt_max = (mem_size / tgt_nr) / size_mb;
+	/*
+	 * Leave a hugepage overhead buffer for DPDK memory management. Reduce
+	 * the DMA upper bound by 200 hugepages per target to account for this.
+	 */
+	bio_chk_cnt_max -= (200 * hugepage_size) / size_mb;
```

Comment:
It'll be broken If hugepage_size is 1GB, and please add an assert (bio_chk_cnt_max > overhead_sz) before doing the minus operation.